File: RUBiS-1.4.3/EJB_EntityBean_id/edu/rice/rubis/beans/servlets/StoreBuyNow.java
package edu.rice.rubis.beans.servlets;

import java.io.IOException;

import javax.naming.Context;
import javax.naming.InitialContext;
import javax.rmi.PortableRemoteObject;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.transaction.UserTransaction;

import edu.rice.rubis.beans.BuyNowHome;
import edu.rice.rubis.beans.Item;
import edu.rice.rubis.beans.ItemHome;
import edu.rice.rubis.beans.ItemPK;
import edu.rice.rubis.beans.TimeManagement;

/** This servlet records a BuyNow in the database and display
 * the result of the transaction.
 * It must be called this way :
 * <pre>
 * http://..../StoreBuyNow?itemId=aa&userId=bb&minBuyNow=cc&maxQty=dd&BuyNow=ee&maxBuyNow=ff&qty=gg 
 *   where: aa is the item id 
 *          bb is the user id
 *          cc is the minimum acceptable BuyNow for this item
 *          dd is the maximum quantity available for this item
 *          ee is the user BuyNow
 *          ff is the maximum BuyNow the user wants
 *          gg is the quantity asked by the user
 * </pre>
 * @author <a href="mailto:cecchet@rice.edu">Emmanuel Cecchet</a> and <a href="mailto:julie.marguerite@inrialpes.fr">Julie Marguerite</a>
 * @version 1.0
 */

public class StoreBuyNow extends HttpServlet
{
  

  private void printError(String errorMsg, ServletPrinter sp)
  {
    sp.printHTMLheader("RUBiS ERROR: StoreBuyNow");
    sp.printHTML("<h2>Your request has not been processed due to the following error :</h2><br>");
    sp.printHTML(errorMsg);
    sp.printHTMLfooter();
  }


  /**
   * Call the <code>doPost</code> method.
   *
   * @param request a <code>HttpServletRequest</code> value
   * @param response a <code>HttpServletResponse</code> value
   * @exception IOException if an error occurs
   * @exception ServletException if an error occurs
   */
  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
  {
    doPost(request, response);
  }

  /**
   * Store the BuyNow to the database and display resulting message.
   *
   * @param request a <code>HttpServletRequest</code> value
   * @param response a <code>HttpServletResponse</code> value
   * @exception IOException if an error occurs
   * @exception ServletException if an error occurs
   */
  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
  {
    ServletPrinter sp = null;
    Context initialContext = null;
    Integer userId; // item id
    Integer itemId; // user id
    float   minBuyNow; // minimum acceptable BuyNow for this item
    float   BuyNow;    // user BuyNow
    float   maxBuyNow; // maximum BuyNow the user wants
    int     maxQty; // maximum quantity available for this item
    int     qty;    // quantity asked by the user

    sp = new ServletPrinter(response, "StoreBuyNow");

    /* Get and check all parameters */

    String value = request.getParameter("userId");
    if ((value == null) || (value.equals("")))
    {
      printError("<h3>You must provide a user identifier !<br></h3>", sp);
      return ;
    }
    else
      userId = new Integer(value);

    value = request.getParameter("itemId");
    if ((value == null) || (value.equals("")))
    {
      printError("<h3>You must provide an item identifier !<br></h3>", sp);
      return ;
    }
    else
      itemId = new Integer(value);


    value = request.getParameter("maxQty");
    if ((value == null) || (value.equals("")))
    {
      printError("<h3>You must provide a maximum quantity !<br></h3>", sp);
      return ;
    }
    else
    {
      Integer foo = new Integer(value);
      maxQty = foo.intValue();
    }

    value = request.getParameter("qty");
    if ((value == null) || (value.equals("")))
    {
      printError("<h3>You must provide a quantity !<br></h3>", sp);
      return ;
    }
    else
    {
      Integer foo = new Integer(value);
      qty = foo.intValue();
    }

    /* Check for invalid values */

    if (qty > maxQty)
    {
      printError("<h3>You cannot request "+qty+" items because only "+maxQty+" are proposed !<br></h3>", sp);
      return ;
    }      

    try
    {
      initialContext = new InitialContext();
    } 
    catch (Exception e) 
    {
      printError("Cannot get initial context for JNDI: " + e+"<br>", sp);
      return ;
    }
    // We want to start transactions from client
    UserTransaction utx = null;
    try
    {
      utx = (javax.transaction.UserTransaction)initialContext.lookup(Config.UserTransaction);
      utx.begin();	
    } 
    catch (Exception e)
    {
      printError("Cannot lookup UserTransaction: "+e+"<br>", sp);
      return ;
    }

    // Try to find the Item corresponding to the Item ID
    Item item;
    try 
    {
      ItemHome itemHome = (ItemHome)PortableRemoteObject.narrow(initialContext.lookup("ItemHome"),
                                                                ItemHome.class);
      item = itemHome.findByPrimaryKey(new ItemPK(itemId));
      item.setQuantity(item.getQuantity() - qty);
      if (item.getQuantity() == 0)
        item.setEndDate(TimeManagement.currentDateToString());
    } 
    catch (Exception e)
    {
      printError("Cannot update Item: " +e+"<br>", sp);
      try
      {
        utx.rollback();
      }
      catch (Exception se) 
      {
        printError("Transaction rollback failed: " + e +"<br>", sp);
      }
      return ;
    }
    try 
    {
      BuyNowHome bHome = (BuyNowHome)PortableRemoteObject.narrow(initialContext.lookup("BuyNowHome"),
                                                      BuyNowHome.class);
      edu.rice.rubis.beans.BuyNow b = bHome.create(userId, itemId, qty);
      utx.commit();
      sp.printHTMLheader("RUBiS: BuyNow result");
      if (qty == 1)
        sp.printHTML("<center><h2>Your have successfully bought this item.</h2></center>\n");
      else
        sp.printHTML("<center><h2>Your have successfully bought these items.</h2></center>\n");
    }
    catch (Exception e)
    {
      printError("Error while storing the BuyNow (got exception: " +e+")<br>", sp);
      try
      {
        utx.rollback();
      }
      catch (Exception se) 
      {
        printError("Transaction rollback failed: " + e +"<br>", sp);
      }
      return ;
    }
		
    sp.printHTMLfooter();
  }

}

File: RUBiS-1.4.3/PHP/StoreBuyNow.php
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <body>
    <?php
    $scriptName = "StoreBuyNow.php";
    include("PHPprinter.php");
    $startTime = getMicroTime();
    
    $userId = $HTTP_POST_VARS['userId'];
    if ($userId == null)
    {
      $userId = $HTTP_GET_VARS['userId'];
      if ($userId == null)
      {
         printError($scriptName, $startTime, "StoreBuyNow", "<h3>You must provide a user identifier!<br></h3>");
         exit();
      }
    }
      
    $itemId = $HTTP_POST_VARS['itemId'];
    if ($itemId == null)
    {
      $itemId = $HTTP_GET_VARS['itemId'];
      if ($itemId == null)
      {
         printError($scriptName, $startTime, "StoreBuyNow", "<h3>You must provide an item identifier !<br></h3>");
         exit();
      }
    }
      
    $maxQty = $HTTP_POST_VARS['maxQty'];
    if ($maxQty == null)
    {
      $maxQty = $HTTP_GET_VARS['maxQty'];
      if ($maxQty == null)
      {
         printError($scriptName, $startTime, "StoreBuyNow", "<h3>You must provide a maximum quantity !<br></h3>");
         exit();
      }
    }

    $qty = $HTTP_POST_VARS['qty'];
    if ($qty == null)
    {
      $qty = $HTTP_GET_VARS['qty'];
      if ($qty == null)
      {
         printError($scriptName, $startTime, "StoreBuyNow", "<h3>You must provide a quantity !<br></h3>");
         exit();
      }
    }

    /* Check for invalid values */

    if ($qty > $maxQty)
    {
      printError("<h3>You cannot request $qty items because only $maxQty are proposed !<br></h3>");
      return ;
    }      

    getDatabaseLink($link);
    beginRW($link);

    if (!$postgres)
      sql_query("LOCK TABLES buy_now WRITE, items WRITE", $link) or die("ERROR: Failed to acquire locks on items and buy_now tables.");
    $result = sql_query("SELECT * FROM items WHERE items.id=$itemId", $link) or die("ERROR: Query failed");
    if (sql_num_rows($result) == 0)
    {
      printError($scriptName, $startTime, "BuyNow", "<h3>Sorry, but this item does not exist.</h3><br>");
      exit();
    }
    $row = sql_fetch_array($result);
    $newQty = $row["quantity"]-$qty;

    $now = virtualTimeSQL();

    txcache_inval("items", "id", $itemId);
    txcache_inval("items", "category", $row["category"]);
    txcache_inval("items", "seller", $row["seller"]);

    if ($newQty == 0)
      sql_query("UPDATE items SET end_date='$now',quantity=$newQty WHERE id=$itemId", $link) or die("ERROR: Failed to update item");
    else
      sql_query("UPDATE items SET quantity=$newQty WHERE id=$itemId", $link) or die("ERROR: Failed to update item");
    // Add BuyNow to database
    $result = sql_query("INSERT INTO buy_now VALUES ($ID_DEFAULT, $userId, $itemId, $qty, \"$now\") RETURNING id", $link) or die("ERROR: Failed to insert new BuyNow in database.");
    $row = sql_fetch_array($result);
    txcache_inval("buy_now", "id", $row["id"]);
    txcache_inval("buy_now", "item_id", $itemId);
    txcache_inval("buy_now", "buyer_id", $userId);
    if (!$postgres)
      sql_query("UNLOCK TABLES", $link) or die("ERROR: Failed to unlock items and buy_now tables.");

    printHTMLheader("RUBiS: BuyNow result");
    if ($qty == 1)
      print("<center><h2>Your have successfully bought this item.</h2></center>\n");
    else
      print("<center><h2>Your have successfully bought these items.</h2></center>\n");
    
    sql_free_result($result);
    commit($link);
    sql_close($link);
    
    printHTMLfooter($scriptName, $startTime);
    ?>
  </body>
</html>
